File: app/api/videos.py
from fastapi import APIRouter, Depends, HTTPException, status

from app.api.auth import require_bearer_token
from app.deps import get_job_service
from app.schemas.job import CreateJobRequest, JobResponse
from app.services.job_service import (
    JobNotCancellableError,
    JobNotFoundError,
    JobService,
)

router = APIRouter(
    prefix="/api/videos",
    tags=["videos"],
    dependencies=[Depends(require_bearer_token)],
)


@router.post("", status_code=status.HTTP_202_ACCEPTED, response_model=JobResponse)
async def create_video(
    req: CreateJobRequest,
    service: JobService = Depends(get_job_service),
) -> JobResponse:
    job = await service.create_job(req)
    return JobResponse.model_validate(job)


@router.get("/{job_id}", response_model=JobResponse)
async def get_video(
    job_id: int,
    service: JobService = Depends(get_job_service),
) -> JobResponse:
    try:
        job = await service.get_job(job_id)
    except JobNotFoundError as e:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(e)) from e
    return JobResponse.model_validate(job)


@router.delete("/{job_id}", response_model=JobResponse)
async def cancel_video(
    job_id: int,
    service: JobService = Depends(get_job_service),
) -> JobResponse:
    try:
        job = await service.cancel_job(job_id)
    except JobNotFoundError as e:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(e)) from e
    except JobNotCancellableError as e:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=str(e)) from e
    return JobResponse.model_validate(job)

File: frontend/src/lib/api/videos.ts
import { apiFetch } from "@/lib/api/client";
import type { CreateVideoRequest, JobResponse } from "@/types/api";

export function createVideo(req: CreateVideoRequest): Promise<JobResponse> {
  return apiFetch<JobResponse>("/api/videos", { method: "POST", body: req });
}

export function getVideo(id: number, signal?: AbortSignal): Promise<JobResponse> {
  return apiFetch<JobResponse>(`/api/videos/${id}`, { signal });
}

export function cancelVideo(id: number): Promise<JobResponse> {
  return apiFetch<JobResponse>(`/api/videos/${id}`, { method: "DELETE" });
}

export function resumeVideo(id: number): Promise<JobResponse> {
  return apiFetch<JobResponse>(`/api/videos/${id}/resume`, { method: "POST" });
}
